### ISSUE TITLE
Relicense request

### ISSUE THREAD

**Issue by sharwell** [gitextensions/gitextensions#4737]
I noticed the project is currently licensed under a GPL license. As a contributor, this license works well for me in the following conditions:

1. The contribution is completely my own code, so I can relicense it if I choose, e.g. #4735
2. The contribution is known throwaway code (nothing particularly new or interesting from an algorithms perspective), e.g. #4601 or #4717 

However, this license is strong barrier to progress on novel work. For example, I will be unable to work with other contributors to address some of the outstanding hard problem related to asynchronous tasks that are discussed in topics like #4501, or work with others to build upon the 

In order to proceed as a participant, one of the following needs to occur:

1. GitExtensions needs to be relicensed under a contributor-friendly open source license¹
2. A utility library needs to be created with GitExtensions which is licensed under a contributor-friendly open source license, with new and interested code related to hard problems placed primarily within that library
3. A new repository created for work related to the above, which is licensed under a contributor-friendly open source license

/cc @drewnoakes @jbialobr @gerhardol @RussKie @KindDragon 

¹ The contributor-friendly licenses I would consider contributing under with other users, in order of most desirable to least, are MIT (with or without an explicit patent grant), Apache 2.0, or BSD (2- or 3-clause).

**Comment by gerhardol** [comment]
I prefer 3

GPL is contributor friendly in the sense that it is not permitted to steal the code for other projects.

**Comment by RussKie** [comment]
I acknowledge the request. In principal I am not opposed to the motion.
I will need to spend some time studying different licensing options.

On 30 Mar 2018 9:57 am, "Gerhard Olsson" <notifications@github.com> wrote:

I prefer 3

GPL is contributor friendly in the sense that it is not permitted to steal
the code for other projects.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/gitextensions/gitextensions/issues/4737#issuecomment-377397434>,
or mute the thread

**Comment by drewnoakes** [comment]
Thanks for bringing this up @sharwell. I would also be happy to see the project under a less viral license. GPL gives me some caution when it comes to contributing.

With so many contributors to the project to date it will probably be very difficult to change the license however.

Does option 3 mean a submodule is not an option? Would it have to be a binary reference? That would suck from a practicality standpoint.

**Comment by sharwell** [comment]
If we can get an agreement from the primary contributors that this is a desirable _direction_, I would be willing to help move this forward by bringing in subject matter experts to help with the process. Given the duration of this project's history and number of contributors, my recommendation¹ would be to approach the .NET Foundation for assistance. From their [About](https://www.dotnetfoundation.org/about) page:

> ### IP and Legal
>
> We help projects organize and manage their IP assignments as well as provide infrastructure for tracking a common, simple, Contribution License Agreement across all .NET Foundation projects. We can help in ensuring people respect the open source licenses and trademarks of projects assigned to the .NET Foundation and we can provide access to advice on legal issues affecting projects in the .NET Foundation.

However, if this is _not_ a desirable direction from the maintainers, I would likely proceed in an independent discussion with @drewnoakes about creating a library aimed at solving some of the problems observed in this project, and then adopting that library in GitExtensions as it becomes ready. A submodule could be appropriate, but likely wouldn't be required.

¹ This is a personal recommendation, fully independent from my current employer.

**Comment by RussKie** [comment]
There are a number of topics, including SO, about re-licensing. It seems
possible to re-license and/or to license under several licences.

**Comment by jbialobr** [comment]
How long can it take to relicense?

**Comment by sharwell** [comment]
> How long can it take to relicense?

I'm not sure. I wouldn't want to approach a legal team about the matter without knowing in advance that the maintainers were generally in favor of attempting to go in that direction.

I will pose it as a question that y'all can vote on: **provided *all* of the follow are met, are you in favor (:+1:) or against (:-1:) the effort?**

1. The new license allows the code and application to be used in strictly more places than before (by choosing a GPL-compatible license or keeping the GPL under a dual license)
2. The expenses, if any, are covered
3. The effort does not require a substantial time investment from maintainers
4. The effort does not make it more difficult to maintain the project

**Comment by drewnoakes** [comment]
I voted with 🎉 because I don't feel I have the same voting rights as others.

**Comment by sharwell** [comment]
@drewnoakes You have the 4th most commits, so it's relevant.

If we can get @spdr870 to :+1: it as well, then I believe it's enough to proceed.

**Comment by gerhardol** [comment]
Relicensing is tricky. Just adding a new project requires no relicensing.
Personally prefer GPL licenses, will accept other licenses.

> Does option 3 mean a submodule is not an option?

It is OK to combine a project with a compatible license with GPL projects as this. Con-Emu is BSD 3-clause for instance. The licenses include the licenses sharwell mentioned, Apache 2.0 seem to be recommended. This project can be added as a submodule. This would require a two step PR for changes. I do not have a problem with that for maintaining.

@drewnoakes 
> I voted with 🎉 because I don't feel I have the same voting rights as others.

Your vote is for sure worth a lot

**Comment by pmiossec** [comment]
Disclaimer : I'm not for or against it but I really want to understand... 

>However, this license (GPL) is strong barrier to progress on novel work.

Why? I really don't get it. 
What is different between GPL and MIT licence in term of contribution?

GPL  guarantee that your work won't be used in another software with another license (proprietary for example) where MIT allows it.

If you have the copyright of the code you include, there is no difference in terms of contribution between GPL and MIT. You also could include MIT code in a gpl software where the opposite is not allowed so if you change the license, you won't be able to include GPL code inside Gitextensions. 

And I just want you to know, that, because no contributors gave their copyright assignment, they still own their copyright. So, if you want to respect the law/license, you **must** get the rights of **all** the contributors to be able to change the license or remove all the code contributed by the ones that refuse or didn't answer (blame all the files and see who still has its code used). 

I don't know if it worth the effort for a software that is not commercialized... But, perhaps, once I understand the "strong barrier"... 

See "copyright assignment" in https://stackoverflow.com/questions/5419923/can-gpl-be-re-licensed

The only gaining I could see is a philosophical one where some potential contributors see GPL as evil and didn't want to contribute to a gpl project...

I like to understand the clear gain here...

**Comment by sharwell** [comment]
> I like to understand the clear gain here...

GitExtensions faces some interesting challenges related to e.g. UI testing and prioritization of work. If I provided a general and reusable solution, that's great and of course I could dual license the contribution so it helps other projects. However, if I want to work with other users, e.g. @drewnoakes who has shown interest in solving similar problems, then it becomes very difficult for us to jointly develop a solution without having it only usable in GitExtensions. I don't want to spend my time creating solutions to hard problems, only to have them be unusable in the real world.

> So, if you want to respect the law/license, ...

[IANAL](https://en.wikipedia.org/wiki/IANAL), so if this moves forward I'm going to let professionals in the industry - lawyers specializing in software licensing - dictate exactly what is and is not required in order for the transition to be successful. Only with that information can the maintainers decide if it's viable.

**Comment by drewnoakes** [comment]
> However, if I want to work with other users, e.g. @drewnoakes who has shown interest in solving similar problems, then it becomes very difficult for us to jointly develop a solution without having it only usable in GitExtensions

Is this because the ability to apply such multi-author contributions elsewhere would require consent from all authors, and you're concerned that might be impractical or not assured?

**Comment by gerhardol** [comment]
The easy path from a legal point of view is to add a new new project, add it as a submodule, and re-implement the parts that need other licenses.

**Comment by RussKie** [comment]
@sharwell it looks like we all in accord. Feel free to engage Legals.

**Comment by RussKie** [comment]
SharpZipLib has gone through a somewhat similar process in the past https://github.com/icsharpcode/SharpZipLib/issues/103

**Comment by jdunne525** [comment]
I am a lowly user like any other and only today started looking at this project on git hub. I must say this all seems a bit fishy. Do you guys have any idea who this sharwell character is? Just be careful you don't make a deal with the devil. I like using this software and it would be a shame to see some legal thing kill the project. Sorry if I'm off base here.

**Comment by RussKie** [comment]
we do

**Comment by RussKie** [comment]
I've been looking at OSS-friendly WinForms UI control libraries, and Syncfusion seems to tick both boxes.
However the current licence appears to disqualify us from their OSS licensing agreement:

> In order to allow Syncfusion to be used in an open-source project we would need to put a Master  License Agreement (MLA) in place. The Master License Agreement would have certain limitations.    
> These limitations include, but are not limited to:   
>   1. There would be a requirement that you make it clear, in a license, to anyone obtaining your > code, that they need to independently obtain dependencies (our product) from us whether under a community, commercial or other license.   
>   2. You will not be able to distribute our binaries as part of the open source project.  
>   3. The project can only be licensed under MIT, Apache, or BSD. The project could not be licensed under GPL or similar licenses. 

I guess I'm happy to attempt to get the ball rolling by sending an email to all those contributed asking whether they'd be happy to relicence.
I would also propose we add a line to our PR template that states something like:

    By contributing this code I acknowledge I have full ownership of the code 
    and I release this contribution under MIT licence.

Thoughts?

**Comment by vbjay** [comment]
Syncfusion:

Oh dear lord.  I tried their stuff out.  I got HOUNDED by them trying to push their sales down my throat. I had to threaten a federal lawsuit to get them to quit.   I'll get you the details later on.

**Comment by RussKie** [comment]
We can require a sign-off and verify it with a Probot - https://github.com/probot/dco

**Comment by sharwell** [comment]
> It requires all commit messages to contain the Signed-off-by line 

This sounds like unnecessary pain. No other project I'm on, including commercial OSS projects, requires this complexity.

**Comment by vbjay** [comment]
![image](https://user-images.githubusercontent.com/3629489/51385427-4218be80-1aed-11e9-9275-ebf4f64c4305.png)

**Comment by vbjay** [comment]
I recommend we add a hook so that when we implement this, it just signs off for us. We can have a githooks folder and do something like https://stackoverflow.com/a/3462998/815813.  We can even add a script that helps setup the symlink. That way, all we need to do is commit.  the signoff just happens.  

We could use something like https://lubomir.github.io/en/2016-05-04-signoff-hooks.html

**Comment by vbjay** [comment]
FYI.  

https://git-scm.com/docs/git-config/1.8.2#git-config-formatsignoff

Also aliasing git commit to run ``` commit -s``` or ``` !git commit -s``` fails.

**Comment by drewnoakes** [comment]
Rather than require sign off in commit messages, which is ambiguous in meaning, we could set up a Contributor License Agreement (CLA).

The .NET Foundation has one, and their projects require contributors to sign the CLA. It's an automated process, carried out the CLA bot. I don't know how difficult this would be to set up, but I think it's worth investigating.

https://cla.dotnetfoundation.org/

**Comment by mrexodia** [comment]
After some confusion it seems like you set up a CLA with the name of a DCO. Please change the text in `contributing.txt` because in my opinion the spirit of what you are doing is not properly reflected in the first few paragraphs of the document. I am not only asserting that I have all the rights to contribute my changes (they are all my own etc etc), but I am also contributing them to you under MIT.

Just let some lawyers write a CLA (@sharwell seems to imply that all expenses will be covered) and link this issue there for context. Now you copied the Linux DCO and inserted some additional changes which looks suspicious to me.

Setting aside the confusion, I think the motivation of the re-license (regardless if everybody agrees or not) is poorly explained. @sharwell asserts that it is for some reason impossible to work with @drewnoakes on a submodule with a permissive license of his choosing and when asked why there seems to be no clear answer. Obviously this point is a bit late, but I love this project and the GPL lover in me doesn't like this one bit.

**Comment by RussKie** [comment]
> I am not only asserting that I have all the rights to contribute my changes (they are all my own etc etc), but I am also contributing them to you under MIT.

Yes, as a contributor you need to assert you have rights to your contribution, and that you release the said changes to us under MIT.


> Now you copied the Linux DCO and inserted some additional changes which looks suspicious to me.

Please read https://opensource.com/article/18/3/cla-vs-dco-whats-difference


> Just let some lawyers write a CLA (@sharwell seems to imply that all expenses will be covered)

It is neither easy nor fast and not in a way you may be imagining it.

> Setting aside the confusion, I think the motivation of the re-license (regardless if everybody agrees or not) is poorly explained

I can agree with this sentiment There were/are a number of discussions both public and private pertaining to the relicensing initiative.

> I love this project and the GPL lover in me doesn't like this one bit.

Which bit? The bit where we ask you to certify the work is yours or the bit where we ask it to release under MIT?

**Comment by mrexodia** [comment]
> Yes, as a contributor you need to assert you have rights to your contribution, and that you release the said changes to us under MIT.

You seem to confuse **you need** with **we would like you to**.

> Please read https://opensource.com/article/18/3/cla-vs-dco-whats-difference

I read it and when looking at https://developercertificate.org/ there are many similarities with your `contributing.txt`, which is why it superficially looked like a simple DCO to me. Notice however that the Linux DCO is only about you having the rights to contribute your changes under the license.

For example section a in the Linux DCO:

> The contribution was created in whole or in part by me and I
> have the right to submit it under the open source license
> indicated in the file; or

Vs `contributing.txt`:

> I am the creator of all my contributions. I am the author of all
> contributed work submitted and further warrant and represent that
> such work is my original creation and I have the right to license
> it to The Git Extensions Project for release under the MIT and/or GPL
> license.

So far it seems similar in spirit, however then there is an additional sentence in `contributing.txt`:

> I hereby grant The Git Extensions Project a nonexclusive,
> irrevocable, royalty-free, worldwide license to reproduce,
> distribute, prepare derivative works, and otherwise use my
> contributions as part of the Git Extensions project, associated
> documentation, books, and tools at no cost to The Git Extensions Project.

This looks very similar to the copyright assignment clause in my employment contract.

Section 2 also seems similar to section a. Section 3 and c are different, but then Section 4 and d seem to match again.

My complaint is not with the contents of your file, just with the wording. For example in section 4/d the phrase `in the file` is used. Which file? I am assuming it means the `LICENSE.md` file which is GPL, however earlier you speak of MIT and this is confusing. It looks like you just forgot to change that wording to `MIT and/or GPL license`?

Perhaps it would help if you show what the `contributing.txt` file was based on. Was this adapted from somewhere? Was it modified? If it was, how did the modifications come to be? I tried finding out for a while, but it looks like there is a random commit adding the file without even a link to this discussion, let alone one where it actually discusses the origin of the file.

I think your intent is for the contributer to write over all their rights to their contribution to you "The Git Extensions Project". You could use for example https://www.docracy.com/0pmrhl5zsdf/individual-contributor-license-agreement-template-cla to make this very clear. Setting aside whether I want to sign it or not it would at least be clear what the intent is, now it takes forever to properly understand what you are signing.

> Which bit? The bit where we ask you to certify the work is yours or the bit where we ask it to release under MIT?

The bit where you want to relicense to MIT. But let's put that aside, since that is already settled. I just think you should reword `contributing.txt` to be a proper CLA since that seems to be the intent. Otherwise you can just change `LICENSE.md` to MIT and paste the linux DCO to keep things simple.

**Comment by RussKie** [comment]
> Otherwise you can just change LICENSE.md to MIT and paste the linux DCO
to keep things simple.

IIUIC, we can't change the licence until all contributors released their
changes to the Project under the MIT licence.

**Comment by RussKie** [comment]
> Perhaps it would help if you show what the `contributing.txt` file was based on. Was this adapted from somewhere? Was it modified? If it was, how did the modifications come to be? I tried finding out for a while, but it looks like there is a random commit adding the file without even a link to this discussion, let alone one where it actually discusses the origin of the file.

It came from the ANTLR project
https://github.com/gitextensions/gitextensions/pull/6124#issuecomment-456574767

**Comment by sharwell** [comment]
The current DCO is an intermediate step so we can review the project history and work towards getting required agreements to change the license to MIT. Once that is complete, we would like to have the project join the .NET Foundation and from then on would use their CLA process.

**Comment by RussKie** [comment]
To update everyone, I have initiated a discussion with the .NET Foundation's legal counsel regarding our options to re-licence the project.
We are also exploring our options in joining .NET Foundation.

**Comment by mwerle** [comment]
Having just received an email asking for my permission, I'd like to add my 2c based on this conversation (I'm not privy to any other discussions which may have happened). Not sure how warranted that is since my own contributions were extremely minor, however:

1. I agree with others above that the reasonings/explanations behind the relicensing are somewhat insubstantial and poorly expanded upon when more details were requested.

2. When I contributed, the project was equally well supported under Linux as well as Windows. That has changed and further development concentrated on Windows only whereas other platforms have been placed in maintenance.  While I don't have time to rectify this myself, and am continuously grateful for the tool irregardless, I was somewhat disappointed by that direction.

3. Expanding on the previous point, an MIT license allows commercialisation. To be blunt, is that the ultimate aim of this relicense?  While in general I have no problems with MIT licenses (and have contributed to other projects under MIT), I most certainly would not be happy if that is the aim.

As I pointed out, my own contributions are so insignificant as to make no difference, so I won't stand in the way of a relicense as long as all major contributors/owners of the project are on board with it, despite my own reservations.  However, I do think all contributors are owed  a full and detailed explanation as to the reasons behind this relicensing.

Taking a software module from a GPL piece of software to other projects (with other licenses) should be no problem as long as all authors which contributed towards that module agree to it.

**Comment by szaliszali** [comment]
Hi, I also arrived here because of the email sent out to all by @RussKie, and I see that there are multiple issues going on in this issue with several years of delay which is rather confusing for me to process. Adds to my confusion that the last relevant update is about joining .NET Foundation, but according to their publicly available FAQ (https://www.dotnetfoundation.org/faq) they would be fine with any OSI-approved open source license (including GPL), so - based on this information - relicensing could be handled separately, or even might be no longer relevant¹.

I don't believe that the current situation that all contributors have to go through that much information is contributor-friendly. Could you please summarize the goal and the current status of the ongoing and/or proposed changes, or maybe open new issues for independent goals [if any]? Thank you.

¹Also, as far as I can see, there are 380 total contributors (according to the About window) and only 74 signatures in contributors.txt (about 20% of the people), I'm concerned that it could still take years to get DCO signed for everyone or replace their past contributions with "DCO-compliant" ones, this seems a rather huge effort which is not guaranteed to be justified by the benefits of the license change ever.

**Comment by sharwell** [comment]
> Expanding on the previous point, an MIT license allows commercialisation. To be blunt, is that the ultimate aim of this relicense?

I think the original post above represents my interested in the relicense quite well. Some of the approaches used for resolving certain hard problems in asynchronous application design and/or testing would work equally well in other applications. I believe knowledge sharing is one of the best ways to improve the quality of software as a whole. Git Extensions serves as a "proving ground" for things like `CancellationTokenSequence`, which can then be adopted by other projects looking for something similar. The possibility of commercialization of concepts like this don't particularly concern me because I believe the original type presented in this repository is complete (no further work necessary to make it better at its job).

Also note that I don't take any of the donation money received by any projects I participate in (and don't believe I ever have in the past).

> ¹Also, as far as I can see, there are 380 total contributors (according to the About window) and only 74 signatures in contributors.txt (about 20% of the people), I'm concerned that it could still take years to get DCO signed for everyone or replace their past contributions with "DCO-compliant" ones, this seems a rather huge effort which is not guaranteed to be justified by the benefits of the license change ever.

We're now working with a legal firm to go through remaining work. My understanding is not all past contributors (defined as users with a commit in the repository history) will need to agree with the relicense. Examples include:

**Comment by mrexodia** [comment]
Just as a preface: I did not make very significant contributions to Git Extensions, but I do use it every day and I run a somewhat successful GPL-licensed project myself. I think what I see happening here is terribly executed and communicated.

> Some of the approaches used for resolving certain hard problems in asynchronous application design and/or testing would work equally well in other applications. I believe knowledge sharing is one of the best ways to improve the quality of software as a whole.

Exactly. This is the reason I feel confident spending a large portion of my time on GPL-licensed projects. The work is interesting, but I also know that nobody will just take it, improve upon it, sell it and not contribute back to the original project.

Quote from the original issue description:

> I will be unable to work with other contributors to address some of the outstanding hard problem related to asynchronous tasks that are discussed in topics like #4501, or work with others to build upon the work in #4735 and make it suitable for use in complicated orchestration sequences.

You could have quite simply put the MIT license header in a comment and write a small note explaining you would like this code to be usable under MIT. As long as the people you worked with agreed this was a fine choice to solve this particular problem (especially since you just said the feature is finished). Instead you seemingly immediately involved lawyers to relicense the whole project (for what now seems to be a finished feature).

Additionally I never saw this issue occur in any project and I do not think it's a common one, but please provide any data to the contrary.

> My understanding is not all past contributors (defined as users with a commit in the repository history) will need to agree with the relicense.

I think you should tread carefully here. Whether you can do something legally and whether it's (morally) justified to do are two completely different things. If you have two holdouts with minor changes I think it'd be justified, but if it is actually true that the majority of contributors do not **agree** with the change, or question the logic of the arguments presented, then why is it happening?

> Adds to my confusion that the last relevant update is about joining .NET Foundation, but according to their publicly available FAQ (https://www.dotnetfoundation.org/faq) they would be fine with any OSI-approved open source license (including GPL), so - based on this information - relicensing could be handled separately, or even might be no longer relevant¹.

From your original reply @sharwell I also got the impression that switching to MIT was somehow a requirement to join the .NET Foundation, it would be good if you could answer if this is actually the case or not.

**Comment by sharwell** [comment]
> I think what I see happening here is terribly executed and communicated.

I'm happy to answer any specific questions, speaking for myself.

> You could have quite simply put the MIT license header in a comment and write a small note explaining you would like this code to be usable under MIT. As long as the people you worked with agreed this was a fine choice to solve this particular problem (especially since you just said the feature is finished).

The paper trail for this type of approach is complicated, to the point that it often doesn't really work in practice. I felt the overhead of keeping track of this was directly undermining my intent working on the project. I filed this issue to gauge interest from other maintainers (each of whom may have their own unique reasons for wanting to relicense). When the other maintainers agreed, I started looking seriously into options that would allow the change to move forward.

> Instead you seemingly immediately involved lawyers to relicense the whole project (for what now seems to be a finished feature).

Lawyers were not contacted prior to reaching unanimous agreement with all current core maintainers plus the original project author. When we did contact lawyers, our request focused simply on ensuring the paper trail for the relicense was complete and trustworthy. Should the project get a new license, it's important that everyone have a clear understanding of what the new license means, and not be encumbered by the use of "questionably-licensed contributions".

> I think you should tread carefully here. Whether you can do something legally and whether it's (morally) justified to do are two completely different things.

Agreed. However, we're intentionally avoiding turning something into an issue where it's not necessary. We're working with users who are not willing to sign the current agreement individually to find a resolution that works for everyone. Personally, I'm much more concerned about cases where authors from long ago are simply unreachable (the project is more than 10 years old), as opposed to cases where users voice objections.

> From your original reply @sharwell I also got the impression that switching to MIT was somehow a requirement to join the .NET Foundation, it would be good if you could answer if this is actually the case or not.

It's actually the other way around. Joining the .NET Foundation gives us access to the legal resources needed to complete the license change to the satisfaction of all users.

**Comment by GintasS** [comment]
@RussKie should we keep this or close it? I think I saw a similar issue/discussion in the discussion about the joining .NET foundation.

**Comment by kimimaru4000** [comment]
I am not a contributor to this project, however I have used GitExtensions for a very long time and believe it is the best Git GUI out there.

It seems the MIT re-licensing is set in stone; however, I'm curious why the Apache 2.0 licensed wasn't considered given the threat of software patents. The MIT license was written before software patents existed, whereas both GPL v3 and Apache 2.0 cover patents to protect users of the software. I would appreciate any insight into this decision.

**Comment by RussKie** [comment]
I think it was simply because .NET was released under MIT license, we followed the same path. I trust Microsoft and their legal counsel have done all the hard work, and decided the MIT license was most appropriate.

**Comment by sharwell** [comment]
> It seems the MIT re-licensing is set in stone; however, I'm curious why the Apache 2.0 licensed wasn't considered given the threat of software patents. The MIT license was written before software patents existed, whereas both GPL v3 and Apache 2.0 cover patents to protect users of the software. I would appreciate any further insight into this decision.

IANAL, and I'm not aware of any cases in the US where a company sued for patent infringement when someone used code they previously published under the MIT license. However, articles I've read on the subject indicate that the MIT license would likely not be a problem for the user of the code in such a case. For example:

https://opensource.com/article/18/3/patent-grant-mit-license

> This is not true; the GPL **does not forbid any use of the software**, including commercial use.

This is not true. The GPL forbids many uses (both commercial and non-commercial) of the software which the MIT license does not forbid. One goal of the authors of things like [`CancellationTokenSequence`](https://github.com/gitextensions/gitextensions/blob/18c9f86773dc128987f9fa353218636f76ca2b76/GitExtUtils/GitUI/CancellationTokenSequence.cs) is the code can be adopted to improve the quality of any other application that may benefit from a similar pattern. This allows developers to focus on other software engineering problems and more efficiently produce applications which are pleasant for users.

**Comment by kimimaru4000** [comment]
> This is not true. The GPL forbids many uses (both commercial and non-commercial) of the software which the MIT license does not forbid. One goal of the authors of things like [`CancellationTokenSequence`](https://github.com/gitextensions/gitextensions/blob/18c9f86773dc128987f9fa353218636f76ca2b76/GitExtUtils/GitUI/CancellationTokenSequence.cs) is the code can be adopted to improve the quality of any other application that may benefit from a similar pattern. This allows developers to focus on other software engineering problems and more efficiently produce applications which are pleasant for users.

The GPL is [free software](https://www.gnu.org/philosophy/free-sw.html#four-freedoms). Free software adheres to the four freedoms. Directly from the FSF's website:

```
The freedom to run the program as you wish, for any purpose (freedom 0).
The freedom to study how the program works, and change it so it does your computing as you wish (freedom 1). Access to the source code is a precondition for this.
The freedom to redistribute copies so you can help others (freedom 2).
The freedom to distribute copies of your modified versions to others (freedom 3). By doing this you can give the whole community a chance to benefit from your changes. Access to the source code is a precondition for this. 
```

Outside of local laws, there is nothing in the GPL preventing the sale or use of the software. If you are referring to proprietary software, that is different from commercial software, which can indeed be FOSS (Red Hat, NextCloud, Qt, etc).

**Comment by pmiossec** [comment]
@sharwell I think you don't really understand what is the real difference between GPL and MIT. I will try to highlight a little here.

> IANAL, and I'm not aware of any cases in the US where a company sued for patent infringement when someone used code they previously published under the MIT license.

Patents has nothing to do with the license. If you break a patent with MIT or GPL code, you could be sued.

>However, articles I've read on the subject indicate that the MIT license would likely not be a problem for the user of the code in such a case.

For users of code (here I understand you mean "user of the application") MIT and GPL has **NO** consequences (other than with GPL you must provide or give access to the source code of your application).

You could do exactly the same thing with your software: give it for free, sell it, ...

> This is not true. The GPL forbids many uses (both commercial and non-commercial) of the software which the MIT license does not forbid.

So your sentence here is false.

**The real difference between MIT and GPL license is in term of source code sharing and binaries linking. Not in term of usage of the application**

With a source code in MIT, you could do whatever you want with this source code without restriction.
So you could take this MIT source code and include it in a proprietary application.
That's why big companies **really like the MIT license** (and push for it) because that allows them to take the code and include it in their product.
And because you are working for Microsoft, you should often hear that this is the best license. And that's also the reason why the dotnet foundation push this license. You don't have to follow which internal product is using a specific source code.
An example is Microsoft that took years ago the source code of the network stack from BSD unix and include it in Windows.

>  One goal of the authors of things like [`CancellationTokenSequence`](https://github.com/gitextensions/gitextensions/blob/18c9f86773dc128987f9fa353218636f76ca2b76/GitExtUtils/GitUI/CancellationTokenSequence.cs) is the code can be adopted to improve the quality of any other application that may benefit from a similar pattern. 

Good example. But we are talking about source code. Not application ;)

If you use source code in GPL or link to binaries compiled from source in GPL, your source code and your application must be in GPL. The goal of the GPL is to say "hey, if you take advantage of my work, I would like to be able to take advantage of yours if it interest me".

A consequence is that you could include some MIT code in a GPL project (and so the code become GPL) where the opposite is not possible.

Personally, I don't think that some source code of GitExtension will taken and used in another application.
Whereas there is an higher probability that someone would want to take the source code to build it's own application and sell it. If this is the case, I would prefer that the person doing that must be obliged  by the GPL to provide the source code than creating a close source application because MIT license allows him to do it.

Big companies are taking advantage of MIT code to build some products and make money with it.
But because we are not making money with our software and because we are not building libraries to be used by other software (where, in this case, MIT is maybe a better license than GPL), I think we, as a project, are more interested by having access to derivative work source code than having our code reused by other non-GPL projects.

Having code in MIT license don't bring value, in term of development, to our project...
And I'm not sure but I hope that dotnet foundation is not enforcing the MIT license for a project to be part of the foundation. 

So, for my contribution to this project, I highly prefer that the code stay in GPL.

**Comment by mrexodia** [comment]
I don’t think there’s any point in protesting anymore. Someone at Microsoft
decided for weird reasons that GitExtensions will be MIT (even though you
could dual license the parts that triggered this discussion) and that’s
what will happen (even though technically not all contributors agree).

**Comment by jbialobr** [comment]
Let me share my point of view. The GPL license is unpragmatic. There is tons of code that is closed and nothing is going to change this. I like the idea of sharing knowledge for free. I think we all ultimately will benefit from it. And saying for free I mean no strings attached like: I will share mine but you will have to share yours. When I give money to a beggar I don't care what he is going to do with that money - it is up to him. The same is with sharing code. I share my code and I don't care who and how will use it. I am happy to share. I use code shared by others and I see that more and more people and companies are sharing their code. I think that GPL is becoming obsolete (at least in my opinion).

**Comment by sharwell** [comment]
> And because you are working for Microsoft, you should often hear that this is the best license.

My statements on licensing have nothing to do with my employer. I joined Microsoft _because_ they agree with my viewpoints, as opposed to joining and then having these viewpoints.

**Comment by kimimaru4000** [comment]
> If you use source code in GPL or link to binaries compiled from source in GPL, your source code and your application must be in GPL. The goal of the GPL is to say "hey, if you take advantage of my work, I would like to be able to take advantage of yours if it interest me".

Based on your answer, I believe you already understand what I'm going to say, but I figured I'd include it for the sake of clarity. The goal of the GPL is **to preserve user freedom**. That is, the freedom for any user to run, share, study, copy, and modify software. Access to the source code is a precondition based on the nature of software consisting of two parts: human-readable source code and a machine-readable binary. The copyleft ensures that those rights are granted to everyone who receives the software, so they in turn can do the same.

**Comment by mpeter50** [comment]
The [LICENSE.MD](https://github.com/gitextensions/gitextensions/blob/v4/LICENSE.md) file still contains the GPL 3 license.
Did I misunderstand something and the relicensing did not go through? From the comments above, this does not seem to be the case.

**Comment by gerhardol** [comment]
There is no relicensing, it is not in focus right now.

**Comment by RussKie** [comment]
We have received sign off's from the majority of our contributors who have
made substantial contributions. There's probably some code that's been
authored under GPL for which we haven't received sign off's.

Why do you ask?

**Comment by mpeter50** [comment]
> Why do you ask?

Nothing important, I just remembered this issue and wanted to check what is the outcome, and I was a bit confused. But now I understand that it is just on hold now.
